File: cli/internal/domain/branch_lifecycle.go
package domain

import (
	"fmt"
	"strconv"
	"strings"
)

// BranchLifecycleTagPrefix is a reserved immutable-tag namespace used as a
// backward-compatible branch lifecycle event log. Older clients preserve
// these refs as ordinary tags instead of rejecting an unknown ref kind.
const BranchLifecycleTagPrefix = "cxt/branch-state/v1/"

type BranchLifecycleState string

const (
	BranchActive   BranchLifecycleState = "active"
	BranchArchived BranchLifecycleState = "archived"
)

type BranchLifecycleEvent struct {
	Branch     string
	Generation uint64
	State      BranchLifecycleState
	Target     ContentHash
	Ref        Ref
}

func parseBranchLifecycleTagName(name string) (BranchLifecycleEvent, bool, error) {
	if !strings.HasPrefix(name, BranchLifecycleTagPrefix) {
		return BranchLifecycleEvent{}, false, nil
	}
	rest := strings.TrimPrefix(name, BranchLifecycleTagPrefix)
	parts := strings.SplitN(rest, "/", 4)
	if len(parts) != 4 || len(parts[0]) != 20 {
		return BranchLifecycleEvent{}, true, fmt.Errorf("%w: malformed branch lifecycle tag %q", ErrInvalidRef, name)
	}
	generation, err := strconv.ParseUint(parts[0], 10, 64)
	if err != nil || generation == 0 || fmt.Sprintf("%020d", generation) != parts[0] {
		return BranchLifecycleEvent{}, true, fmt.Errorf("%w: invalid branch lifecycle generation in %q", ErrInvalidRef, name)
	}
	state := BranchLifecycleState(parts[1])
	if state != BranchActive && state != BranchArchived {
		return BranchLifecycleEvent{}, true, fmt.Errorf("%w: invalid branch lifecycle state in %q", ErrInvalidRef, name)
	}
	if len(parts[2]) != 64 || parts[2] != strings.ToLower(parts[2]) {
		return BranchLifecycleEvent{}, true, fmt.Errorf("%w: invalid branch lifecycle target in %q", ErrInvalidRef, name)
	}
	target := ContentHash("sha256:" + parts[2])
	if err := ValidateContentHash(target); err != nil {
		return BranchLifecycleEvent{}, true, err
	}
	if err := ValidateBranchName(parts[3]); err != nil {
		return BranchLifecycleEvent{}, true, err
	}
	return BranchLifecycleEvent{Branch: parts[3], Generation: generation, State: state, Target: target}, true, nil
}

func NewBranchLifecycleRef(repoID, branch string, target ContentHash, generation uint64, state BranchLifecycleState) (Ref, error) {
	if err := ValidateBranchName(branch); err != nil {
		return Ref{}, err
	}
	if err := ValidateContentHash(target); err != nil {
		return Ref{}, err
	}
	if generation == 0 || (state != BranchActive && state != BranchArchived) {
		return Ref{}, fmt.Errorf("%w: invalid branch lifecycle event", ErrInvalidRef)
	}
	name := fmt.Sprintf("%s%020d/%s/%s/%s", BranchLifecycleTagPrefix, generation, state, strings.TrimPrefix(string(target), "sha256:"), branch)
	ref := Ref{Kind: RefTag, Name: name, RepoID: repoID, Target: target}
	if err := ValidateRef(ref); err != nil {
		return Ref{}, err
	}
	return ref, nil
}

func ParseBranchLifecycleRef(ref Ref) (BranchLifecycleEvent, bool, error) {
	if ref.Kind != RefTag {
		return BranchLifecycleEvent{}, false, nil
	}
	event, ok, err := parseBranchLifecycleTagName(ref.Name)
	if err != nil || !ok {
		return BranchLifecycleEvent{}, ok, err
	}
	if ref.Target != event.Target {
		return BranchLifecycleEvent{}, true, fmt.Errorf("%w: lifecycle tag target does not match its name", ErrInvalidRef)
	}
	event.Ref = ref
	return event, true, nil
}

func branchLifecycleLater(left, right BranchLifecycleEvent) bool {
	if left.Generation != right.Generation {
		return left.Generation > right.Generation
	}
	if left.State != right.State {
		return left.State == BranchActive
	}
	if left.Target != right.Target {
		return left.Target > right.Target
	}
	return left.Ref.Name > right.Ref.Name
}

func LatestBranchLifecycle(refs []Ref, branch string) (BranchLifecycleEvent, bool, error) {
	var latest BranchLifecycleEvent
	found := false
	for _, ref := range refs {
		event, ok, err := ParseBranchLifecycleRef(ref)
		if err != nil {
			return BranchLifecycleEvent{}, false, err
		}
		if !ok || event.Branch != branch {
			continue
		}
		if !found || branchLifecycleLater(event, latest) {
			latest, found = event, true
		}
	}
	return latest, found, nil
}

// BranchLifecycleStates projects the append-only event refs into one latest
// state per branch in a single pass. List/manifest consumers use this instead
// of rescanning the complete immutable event history for every branch.
func BranchLifecycleStates(refs []Ref) (map[string]BranchLifecycleEvent, error) {
	states := make(map[string]BranchLifecycleEvent)
	for _, ref := range refs {
		event, ok, err := ParseBranchLifecycleRef(ref)
		if err != nil {
			return nil, err
		}
		if !ok {
			continue
		}
		if current, found := states[event.Branch]; !found || branchLifecycleLater(event, current) {
			states[event.Branch] = event
		}
	}
	return states, nil
}

// ProjectBranchLifecycleRefs derives the externally visible mutable
// projections from raw refs plus immutable lifecycle events. An archive hides
// only the exact branch target it observed. If a crash left symbolic HEAD
// pointing at that hidden branch, HEAD is exposed as detached at the archived
// snapshot so manifests never contain a dangling symbolic ref.
func ProjectBranchLifecycleRefs(refs []Ref) ([]Ref, error) {
	states, err := BranchLifecycleStates(refs)
	if err != nil {
		return nil, err
	}
	branches := make(map[string]Ref)
	for _, ref := range refs {
		if ref.Kind == RefBranch {
			branches[ref.Name] = ref
		}
	}
	out := make([]Ref, 0, len(refs))
	for _, ref := range refs {
		if ref.Kind == RefBranch {
			latest, ok := states[ref.Name]
			if ok && latest.State == BranchArchived && latest.Target == ref.Target {
				continue
			}
		}
		if ref.Kind == RefHEAD && ref.Symbolic != "" {
			branch := strings.TrimPrefix(ref.Symbolic, "refs/heads/")
			latest, ok := states[branch]
			raw, exists := branches[branch]
			if ok && latest.State == BranchArchived && (!exists || raw.Target == latest.Target) {
				ref.Symbolic = ""
				ref.Target = latest.Target
			}
		}
		out = append(out, ref)
	}
	return out, nil
}

func NextBranchLifecycleGeneration(refs []Ref, branch string) (uint64, error) {
	latest, ok, err := LatestBranchLifecycle(refs, branch)
	if err != nil {
		return 0, err
	}
	if !ok {
		return 1, nil
	}
	if latest.Generation == ^uint64(0) {
		return 0, fmt.Errorf("%w: branch lifecycle generation exhausted", ErrInvalidRef)
	}
	return latest.Generation + 1, nil
}

File: frontend/web/src/branchLifecycle.ts
import type { Ref } from './types';

export const BRANCH_LIFECYCLE_TAG_PREFIX = 'cxt/branch-state/v1/';

export interface BranchLifecycleEvent {
  branch: string;
  generation: bigint;
  state: 'active' | 'archived';
  target: string;
  ref: Ref;
}

export function parseBranchLifecycleRef(ref: Ref): BranchLifecycleEvent | null {
  if (ref.kind !== 'tag' || !ref.name.startsWith(BRANCH_LIFECYCLE_TAG_PREFIX)) return null;
  const rest = ref.name.slice(BRANCH_LIFECYCLE_TAG_PREFIX.length);
  const match = rest.match(/^([0-9]{20})\/(active|archived)\/([0-9a-f]{64})\/(.+)$/);
  if (!match) return null;
  const target = `sha256:${match[3]}`;
  if (ref.target !== target || !match[4]) return null;
  const generation = BigInt(match[1]);
  if (generation === 0n || generation.toString().padStart(20, '0') !== match[1]) return null;
  return { branch: match[4], generation, state: match[2] as 'active' | 'archived', target, ref };
}

function later(left: BranchLifecycleEvent, right: BranchLifecycleEvent): boolean {
  if (left.generation !== right.generation) return left.generation > right.generation;
  if (left.state !== right.state) return left.state === 'active';
  if (left.target !== right.target) return left.target > right.target;
  return left.ref.name > right.ref.name;
}

export function latestBranchLifecycle(refs: Ref[], branch: string): BranchLifecycleEvent | null {
  let latest: BranchLifecycleEvent | null = null;
  for (const ref of refs) {
    const event = parseBranchLifecycleRef(ref);
    if (!event || event.branch !== branch) continue;
    if (!latest || later(event, latest)) latest = event;
  }
  return latest;
}

export function branchLifecycleStates(refs: Ref[]): Map<string, BranchLifecycleEvent> {
  const states = new Map<string, BranchLifecycleEvent>();
  for (const ref of refs) {
    const event = parseBranchLifecycleRef(ref);
    if (!event) continue;
    const current = states.get(event.branch);
    if (!current || later(event, current)) states.set(event.branch, event);
  }
  return states;
}

/** Projects internal lifecycle events into the active branch list while
 * retaining the events themselves for reachability and replica convergence. */
export function projectBranchRefs(refs: Ref[]): Ref[] {
  const states = branchLifecycleStates(refs);
  const branches = new Map(refs.filter((ref) => ref.kind === 'branch').map((ref) => [ref.name, ref]));
  const out: Ref[] = [];
  for (const input of refs) {
    const ref = { ...input };
    if (ref.kind === 'branch') {
      const latest = states.get(ref.name);
      if (latest?.state === 'archived' && latest.target === ref.target) continue;
    }
    if (ref.kind === 'head' && ref.symbolic) {
      const branch = ref.symbolic.replace(/^refs\/heads\//, '');
      const latest = states.get(branch);
      const raw = branches.get(branch);
      if (latest?.state === 'archived' && (!raw || raw.target === latest.target)) {
        ref.symbolic = '';
        ref.target = latest.target;
      }
    }
    out.push(ref);
  }
  return out;
}

export function archivedBranchMarkers(refs: Ref[]): { branch: string; target: string }[] {
  const projected = projectBranchRefs(refs);
  const active = new Set(projected.filter((ref) => ref.kind === 'branch').map((ref) => ref.name));
  return [...branchLifecycleStates(refs).values()]
    .filter((event) => event.state === 'archived' && !active.has(event.branch))
    .map((event) => ({ branch: event.branch, target: event.target }))
    .sort((left, right) => left.branch.localeCompare(right.branch));
}
